open_spiel/games/maedn.h

```diff
@@ -0,0 +1,281 @@
+  { 12, 7 }, { 12, 8 }, { 12, 9 },  { 12, 10 }, { 10, 10 },
+  { 8, 10 }, { 8, 9 }, { 8, 8 },  { 8, 7 }, { 8, 6 },
+  { 6, 6 }, { 4, 6 }, { 2, 6 },  { 0, 6 }, { 0, 5 },
+  // Goal fields.
+  { 2, 5 }, { 4, 5 }, { 6, 5 }, { 8, 5 },
+  { 10, 1 }, { 10, 2 }, { 10, 3 }, { 10, 4 },
+  { 18, 5 }, { 16, 5 }, { 14, 5 }, { 12, 5 },
+  { 10, 9 }, { 10, 8 }, { 10, 7 }, { 10, 6 },
+  // Off the board fields.
+  { 0, 0 }, { 2, 0 }, { 2, 1 }, { 0, 1 },
+  { 18, 0 }, { 20, 0 }, { 20, 1 }, { 18, 1 },
+  { 18, 10 }, { 20, 10 }, { 20, 9 }, { 18, 9 },
+  { 0, 10 }, { 2, 10 }, { 2, 9 }, { 0, 9 },
+};
+
+// This is a small helper to track historical turn info not stored in the moves.
+// It is only needed for proper implementation of Undo.
+struct TurnHistoryInfo {
+  int player;
+  int prev_player;
+  int dice;
+  int prev_dice;
+  Action action;
+  int thrown_out_player;
+  TurnHistoryInfo(int _player, int _prev_player, 
+                  int _dice, int _prev_dice,
+                  int _action, int _thrown_out_player)
+      : player(_player),
+        prev_player(_prev_player),
+        dice(_dice),
+        prev_dice(_prev_dice),
+        thrown_out_player(_thrown_out_player),
+        action(_action) {}
+};
+
+class MaednGame;
+
+class MaednState : public State {
+ public:
+  MaednState(const MaednState&) = default;
+  MaednState(std::shared_ptr<const Game>, bool two_players_opposite);
+
+  Player CurrentPlayer() const override;
+  void UndoAction(Player player, Action action) override;
+  std::vector<Action> LegalActions() const override;
+  std::string ActionToString(Player player, Action move_id) const override;
+  std::vector<std::pair<Action, double>> ChanceOutcomes() const override;
+  std::string ToString() const override;
+  bool IsTerminal() const override;
+  std::vector<double> Returns() const override;
+  std::string ObservationString(Player player) const override;
+  void ObservationTensor(Player player,
+                         absl::Span<float> values) const override;
+  std::unique_ptr<State> Clone() const override;
+
+  // Setter function used for debugging and tests.
+  // History is not set by this method, so calls to UndoAction will cause
+  // undefined behaviour!
+  void SetState(int cur_player,
+                int prev_player,
+                int dice,
+                int prev_dice,
+                const std::vector<int>& board,
+                const std::vector<int>& out);
+  // Setter function similar to SetState, used to test ObservationTensor.
+  // Some values are not part of ObservationTensor (like prev_player_ and 
+  // prev_dice_) and so have to be given from outside. History is not part
+  // of ObservationTensor either, so calls to UndoAction will cause undefined
+  // behaviour!
+  void FromObservationTensor(Player player, 
+                             absl::Span<float> values, 
+                             Player prev_player,
+                             int prev_dice);
+
+  int dice() const { return dice_; }
```

Comment:
don't see this used anywhere, am I missing it?

Reply:
Yes, it's not needed, it was in the game i used as a template, so i thought it would be nice to have.